python/pyspark/sql/pandas/_typing/protocols/series.pyi

```diff
@@ -0,0 +1,253 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+# This Protocol resuses core Pandas annotation.
+# Overall pipeline looks as follows
+# - Stubgen pandas.core.series
+# - Add Protocol as a base class
+# - Replace imports with Any
+
+import numpy as np  # type: ignore[import]
+from typing import Any, Callable, Hashable, IO, Optional
+from typing_extensions import Protocol
+
+groupby_generic = Any
+
+class SeriesLike(Protocol):
```

Comment:
This is kind of said that we have to do pandas' series in PySpark. Can we avoid?

Reply:
Oh... That's a tricky question. Let me provide some context...

In general this protocols exist not so much to annotate Pandas interfaces, but rather to make our side more robust. Initially, I just type ignored Pandas and was pretty satisfied about it.

However, I missed that on the way, it resulted in a situation, where ignore swallowed all the errors. Since Pandas-related are quite complex and it is easy to make a mistake I introduced this thingy.

So what can be done here:

- Removing protocols and going back to `type: ignore`. To be honest, I am not very fond of that idea, but I see why you're concerned.
- Creating a transient package that will fill the hole we have here.
- Trying to convince good Pandas folks to take a leap [here](https://github.com/pandas-dev/pandas/issues/28142) :)

Reply:
One argument for keeping this as-is, is that it is primarily an internal type checking feature. It is not exposed in a way that can be used by the final user (though I thought about it in the past). 

Thought it might be an argument against keeping it in the first place.

One more option is to extract both Pandas annotations from source just for the use in CI pipeline, until typed version is released. This will ensure consistency on our side.